app/grandchallenge/evaluation/views/api.py

```diff
@@ -23,3 +45,88 @@ class EvaluationViewSet(ReadOnlyModelViewSet):
         *api_settings.DEFAULT_RENDERER_CLASSES,
         PaginatedCSVRenderer,
     )
+
+    @action(detail=True, permission_classes=[CanClaimEvaluation])
+    def claim(self, request, *args, **kwargs):
+        evaluation = self.get_object()
+
+        if not evaluation.status == Evaluation.PENDING:
+            return Response(
+                {"status": "You can only claim pending evaluations."},
+                status=400,
+            )
+
+        evaluation.status = Evaluation.EXECUTING
+        evaluation.started_at = now()
+        evaluation.save()
+
+        serializer = ExternalEvaluationSerializer(
+            instance=evaluation, context={"request": request}
+        )
+        return Response(serializer.data)
+
+    @action(detail=False, methods=["GET"])
+    def claimable_evaluations(self, request, *args, **kwargs):
+        queryset = get_objects_for_user(
+            request.user, "evaluation.claim_evaluation"
+        ).filter(status=Evaluation.PENDING)
+        serializer = ExternalEvaluationSerializer(
+            queryset, many=True, context={"request": request}
+        )
+        return Response(serializer.data)
+
+    @action(
+        detail=True,
+        methods=["PATCH"],
+        permission_classes=[CanClaimEvaluation],
+        serializer_class=ExternalEvaluationUpdateSerializer,
+    )
+    def update_external_evaluation(self, request, *args, **kwargs):
+        evaluation = self.get_object()
+        if evaluation.status != Evaluation.EXECUTING:
```

Comment:
1. How do we differentiate between evaluations that GC is executing and evaluations that are being evaluated externally? This is kindof important for the CloudWatch metrics where we track how busy the system is, so that logic will need to change too.

2. Do you check to ensure that only the externally evaluated algorithms can be patched here? We need to ensure that things we are evaluating cannot be change, and that the entity who is evaluating the submission is the only one who can update.

3. Do we track who claimed the job? Maybe this is needed to ensure the correct entity is updating it again.

4. Is there a timeout on the claim? We should probably track this and mark it as failed if there do not return in time.

Reply:
> 1. How do we differentiate between evaluations that GC is executing and evaluations that are being evaluated externally?

Ok, in that case, I suggest adding a status `CLAIMED` to the option of status choices and using that (set when evaluation is claimed).

Reply:
> 2. Do you check to ensure that only the externally evaluated algorithms can be patched here?

Yes, the `CanClaimEvaluation` object permission checks that the requesting user has the `claim_evaluation` permission for the given evaluation. And that permission is granted when an evaluation is created for an external phase, never for internal evaluations.

Reply:
> 3. Do we track who claimed the job? Maybe this is needed to ensure the correct entity is updating it again.

I need to add that indeed, or else we have no way of checking that the same entity is updating the evaluation.

Reply:
> 4\. Is there a timeout on the claim?

Yes, that is still missing. I was thinking of adding a period task that marks evals as Cancelled if they have been in the claimed state for longer than X hours. We can start with a timeout of 1 day, and change that if necessary.

Reply:
RE: 3 and 4

This is why I proposed using the `nonce` approach like we do with email validation. Sets a token that has a timeout and ensures that it is invalidated once it is used, and has a timeout. Double checks that the correct evaluation is being updated by the correct entity that requested it. If you issue an `nonce` token then you would not need to encode who claimed the evaluation, but can be sure that the evaluation that is being changed is the correct one.

Reply:
I have added a nonce token that is created when an evaluation is claimed, stored on the evaluation instance and checked on update.

Reply:
I have added a `claimed_by` field that is set when the evaluation is claimed and checked, together with the `started_at` time, on update. For now, I set the timeout to 1 day.

Reply:
Wait, just realized that I forgot to check that the requesting user on update doesn't have any other claimed evals.

Reply:
Think this is all solved now! Looks good, just some minor stuff on approach.